packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -236,128 +165,168 @@ fn read_file(path: &str) -> Result<String> {
+        let exception = module.get_exception();
         let exception_string = exception
-            .to_string(&mut try_catch)
+            .to_string(&mut try_catch_scope)
             .unwrap()
-            .to_rust_string_lossy(&mut try_catch);
+            .to_rust_string_lossy(&mut try_catch_scope);
 
-        Error::Eval(path.into(), format!("Execution error: {exception_string}"))
-    })
+        return Err(Error::PackageImport(
+            path.to_string(),
+            format!("Could not evaluate code for package: {exception_string}"),
+        ));
+    }
+
+    module_map.borrow_mut().path_to_module.insert(
+        path.to_string(),
+        v8::Global::new(&mut try_catch_scope, module),
+    );
+
+    Ok(module)
 }
 
-fn import_file<'s>(
-    scope: &mut v8::HandleScope<'s>,
+/// Callback called for each `import ...` in JS files. It reads the file, compiles the source
+/// code, and evaluates it. Modules are only compiled and evaluated once.
+// Simple example without any caching: https://gist.github.com/surusek/4c05e4dcac6b82d18a1a28e6742fc23e
+// More elaborate example with caching and multiple types of imports: https://github.com/denoland/deno/blob/f7e7f548499eff8d2df0872d1340ddcdfa028c45/core/bindings.rs#L1344
+fn module_resolve_callback<'s>(
     context: v8::Local<'s, v8::Context>,
-    path: &str,
-    args: &[Value<'s>],
-) -> Result<Value<'s>> {
-    let v = eval_file(scope, path)?;
-    let f: Function<'_> = v.try_into().map_err(|err| {
-        Error::FileImport(
-            path.into(),
-            format!("Could not convert file into Function: {err}"),
-        )
-    })?;
+    // path of the module trying to get imported
+    specifier: v8::Local<'s, v8::String>,
+    _import_assertions: v8::Local<'s, v8::FixedArray>,
+    _referrer: v8::Local<'s, v8::Module>,
+) -> Option<v8::Local<'s, v8::Module>> {
+    // SAFETY: we are in a callback
+    let mut scope = unsafe { v8::CallbackScope::new(context) };
+    let specifier = specifier.to_rust_string_lossy(&mut scope);
+
+    match import_module(&mut scope, &specifier) {
+        Ok(module) => Some(module),
+        Err(err) => {
+            tracing::error!("Couldn't import {specifier}, {err}.");
 
-    let global_context = context.global(scope);
-    let imported = call_js_function(scope, f, global_context.into(), args)
-        .map_err(|err| Error::FileImport(path.into(), format!("Could not call function: {err}")))?;
-
-    Ok(imported)
+            None
+        }
+    }
 }
 
 impl<'s> Embedded<'s> {
-    fn import(
-        scope: &mut v8::HandleScope<'s>,
-        context: v8::Local<'s, v8::Context>,
-    ) -> Result<Self> {
-        let arrow = eval_file(
-            scope,
-            "./src/worker/runner/javascript/apache-arrow-bundle.js",
-        )?;
+    fn import_common_js_files(scope: &mut v8::HandleScope<'s>) -> Result<Self> {
+        // `hash_stdlib` can't be imported as a module because it needs to be available globally for
+        // behaviors.
```

Comment:
Added a TODO and created an assana task https://github.com/hashintel/hash/pull/493/commits/83e31e88aeaa6f1f7b59eeabe45413acaf5dd959 [Asana task](https://app.asana.com/0/1199548034582004/1202225025969133/f)

One thing I want to note is that users didn't have to care about where the js file were in our codebase. That also mean moving it will become a breaking change for us. Maybe JS can do something to avoid that, I don't know. (I guess we could fake a specific path)

Reply:
I wasn't suggesting users should add that. I'm suggesting _we_ could automatically insert it into the source-code text when we load the behaviour

Reply:
Oooh! sorry I misunderstood